broadinstitute/single_cell_portal_core
# Additional integration for AnnData file upload & parsing (SCP-4957)

#### BACKGROUND & CHANGES
This update adds integration to support uploading, parsing, and updating single `AnnData` files in a study that is leveraging the "AnnData" upload experience created in #1733.  Users can now upload an `AnnData` file using this new view and have all form data sent along with the file and persisted in the database.  This data can then be updated later by using the same forms.  

Parameters necessary for ingest processes are now saved in the `AnnDataFileInfo` object under the `data_fragments` array.  This is an array of objects that mimic `StudyFile` entries to a degree in that they share some of the same attribute names, and can represent the data from expression/metadata/clustering upload forms.  These objects have validation rules to ensure that they contain the required data to avoid saving bad entries.

In addition, this update contains various refactoring of necessary React components, and fixes/enhancements regarding parameter validation & sanitization in associated libraries (such as `Api::V1::StudyFilesController` and `AnnDataIngestParameters`).

#### MANUAL TESTING
1. Boot as normal, including `Delayed::Job` if not using Dockerized setup
2. Sign in and create a new study
3. Select the `AnnData` upload UX
4. Using the `trimmed_compliant_pbmc3K.h5ad` from [scp-ingest-pipeline](https://github.com/broadinstitute/scp-ingest-pipeline/blob/development/tests/data/anndata/trimmed_compliant_pbmc3K.h5ad) source, fill out the forms with the following data:
    * Expression: select any values you like as they are not used yet, this is more for persistence checking
    * Clustering: 
      * `name`: `UMAP`
      * `obsm_key_name`: `X_umap`
5. On the final `AnnData` tab, upload the file and click `Save`
6. Confirm the file uploads and pushes to the bucket, and that you see a message similar to the following in `development.log`:
```
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 6410b66ab3be060055827c6b
  - "--study-file-id"
  - 6410b6a9a87b3bbd53fbc351
  - "--user-metrics-uuid"
  - 0fc4d8e5-6c4c-422e-94a3-c78801757202
  - ingest_anndata
  - "--ingest-anndata"
  - "--anndata-file"
  - gs://fc-2406ca86-4751-4ba7-b331-7fb3300d350f/trimmed_compliant_pbmc3K.h5ad
  - "--extract"
  - '["cluster", "metadata"]'
  - "--obsm-keys"
  - '["X_umap"]'
  ...
```
7. Back in the form, go to the `Expression` tab and enter a value for the `description` and save, confirming that there is no error
8. In a separate terminal window, enter the Rails console and confirm that the update exists in the database:
```
study_file = StudyFile.last
=> #<StudyFile _id: 6410b6a9a87b3bbd53fbc351, created_at: ...>
expression_fragment = study_file.ann_data_file_info.fragments_by_type(:expression).first
=> 
{"_id"=>"6410b688e161e55bb0b2064c",                    
...
expression_fragment[:description]
=> <whatever you entered in the form above>
```
